src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
> That's not a safe assumption for non swing /awt events.

Right, but I bet listeners get ontology change events and then fiddle with the Swing UI without doing any checking at all.

> On the other hand so many things in Protégé that shouldn't be on the EDT are, so it's cosmic
balance :)

Eugh, tell me about it.